kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::LinkedList, sync::Arc};
+use core::{
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+
+use aster_frame::sync::{guard_lock, MutexGuard, SpinLock};
+
+use super::Pauser;
+use crate::Errno;
+
+/// The lock error type for situations where the condvar is not notified.
+pub enum LockErr<Guard> {
+    Timeout(Guard),
+    Interrupted(Guard),
+    Unknown(Guard),
+}
+/// LockResult, different from Rust std.
+/// The result of a lock operation.
+pub type LockResult<Guard> = Result<Guard, LockErr<Guard>>;
+
+impl<Guard> LockErr<Guard> {
+    pub fn unwrap(self) -> Guard {
+        match self {
+            LockErr::Timeout(guard) => guard,
+            LockErr::Interrupted(guard) => guard,
+            LockErr::Unknown(guard) => guard,
+        }
+    }
+}
+
+/// A condition variable.
+pub struct Condvar {
+    pauser: Arc<Pauser>,
+    should_wait: SpinLock<LinkedList<Arc<AtomicBool>>>,
+}
+
+impl Condvar {
+    /// Creates a new condition variable.
+    pub fn new() -> Self {
+        Condvar {
+            pauser: Pauser::new(),
+            should_wait: SpinLock::new(LinkedList::new()),
+        }
+    }
+
+    /// Waits for the condition variable to be signaled or broadcasted.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> LockResult<MutexGuard<'a, T>> {
+        let lock = guard_lock(&guard);
+        drop(guard);
+        self.should_wait
+            .lock()
+            .push_back(Arc::new(AtomicBool::new(true)));
+        let cond = || {
+            let wait_list = self.should_wait.lock();
+            if let Some(wait) = wait_list.front() {
+                if wait.load(Ordering::SeqCst) {
```

Comment:
The statement will cause race condition. Suppose two threads are waiting on the same `CondVar`, then the wait list has two `AtomicBool`, the front one for the first thread and the back one for the second thread. If the second thread is woken up at first, then it will use the front `AtomicBool`, but the `AtomicBool` belongs to the first thread!

I suggest you can change the condition as
```rust
let waiter = Arc::new(AtomicBool::new(true))
self.should_wait.lock().push_back(waiter.clone());
let cond = || {
    if waiter.load(Ordering::SeqCst) {
        Some(())
    } else {
        None
    }
}

```